src/controller/java/src/chip/onboardingpayload/Verhoeff10.kt

```
@@ -0,0 +1,91 @@
+      2, 8, 9, 5, 6, 7, 4, 0, 1, 2, 3, 9, 5, 6, 7, 8, 5, 9, 8, 7, 6, 0, 4, 3, 2, 1, 6, 5, 9, 8, 7, 1, 0, 4,
+      3, 2, 7, 6, 5, 9, 8, 2, 1, 0, 4, 3, 8, 7, 6, 5, 9, 3, 2, 1, 0, 4, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0
+    )
+
+    // Permutation table for the algorithm.
+    val sPermTable = byteArrayOf(1, 5, 7, 6, 2, 8, 3, 0, 9, 4)    
+
+    // Compute a check character for a given string.
+    fun computeCheckChar(str: String): Char {
+      return computeCheckChar(str, str.length)
+    }
+
+    fun computeCheckChar(str: String, strLen: Int): Char {
+      var c = 0
+      for (i in 1..strLen) {
+        val ch = str[strLen - i]
+        val value = charToVal(ch)
+        if (value < 0)
+          return 0.toChar() // invalid character
+        val p = Verhoeff.permute(value, sPermTable, Base, i)
+        c = sMultiplyTable[c * Base + p]
+      }
+      c = Verhoeff.dihedralInvert(c, PolygonSize)
+      return valToChar(c)
+    }
+
+    // Verify a check character against a given string.
+    fun validateCheckChar(checkChar: Char, str: String): Boolean {
+      return validateCheckChar(checkChar, str, str.length)
+    }
+
+    fun validateCheckChar(checkChar: Char, str: String, strLen: Int): Boolean {
+      return computeCheckChar(str, strLen) == checkChar
+    }
+
+    // Verify a check character at the end of a given string.
+    fun validateCheckChar(str: String): Boolean {
+      return validateCheckChar(str, str.length)
+    }
+
+    fun validateCheckChar(str: String, strLen: Int): Boolean {
+      if (strLen == 0)
+        return false
+      return validateCheckChar(str[strLen - 1], str, strLen - 1)
+    }
+
+    // Convert between a character and its corresponding value.
+    fun charToVal(ch: Char): Int {
+      if (ch in '0'..'9')
+        return ch - '0'
+      return -1
+    }
+
+    private fun valToChar(value: Int): Char {
+      if (value in 0..Base)
+        return ('0'.code + value).toChar()
+      return 0.toChar()
```

s/0.toChar/'0'/ ?

---

'0' is not same as 0.toChar(), throw exception instead of return 0